gitutil.py

```diff
@@ -0,0 +1,137 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import subprocess
+
+import git
+
+from util import ensure_not_empty, ensure_directory_exists, fail
+
+def update_submodule(
+    repo_path: str,
+    tree_ish: str,
+    submodule_path: str,
+    commit_hash: str,
+    author: str,
+    email: str,
+):
+    '''Update the submodule of a git-repository to a specific commit.
+
+    Create a new commit, with the passed tree-ish as parent, in the given repository.
+
+    Parameters
+    ------
+    repo_path : str
+        Path to a directory containing an intialised git-repo with a submodule to update.
+    tree_ish : str
+        Valid tree-ish to use as base for creating the new commit. This will be used as parent for the commit created by this function.
+        Example: 'master' for the head of the master-branch.
+    submodule_path : str
+        Path (relative to `repo_path`) to the submodule. Must be immediately below the root (i.e. `repo_path`) of the repository.
+    commit_hash : str
+        The hash the submodule should point to in the created commit. This should be a valid commit hash in the submodule's repository.
+    author : str,
+        Will be set as author of the created commit
+    email : str
+        Will be set for the author of the created commit
+
+    Returns
+    ------
+    str
+        The hexadecimal SHA-1 hash of the created commit
+    '''
+    ensure_directory_exists(repo_path)
+    repo_path = os.path.abspath(repo_path)
+
+    ensure_not_empty(submodule_path)
+    if '/' in submodule_path:
+        fail('Specified submodule-path is not in the top-level of the repository: {s}'.format(s=submodule_path))
+
+    ensure_not_empty(tree_ish)
+    ensure_not_empty(commit_hash)
+    ensure_not_empty(author)
+    ensure_not_empty(email)
+
+    repo = git.Repo(repo_path)
+
+    # Fetch tree-ish from git.repo and create mktree-parseable string-representation of it.
+    # Since submodule-objects contain the SHA of the to-be-checked-out version of the
+    # target repository, we replace that during the string building.
+    tree = repo.tree(tree_ish)
+    tree_representation = '\n'.join(_serialise_and_patch_tree(tree, submodule_path, commit_hash))
+
+    # Pass the serialised tree to git mk-tree using GitPython. We cannot do this in GitPython
+    # directly as it does not support arbitrary tree manipulation.
+    auto_interrupt = repo.git.mktree(istream = subprocess.PIPE, as_process=True)
```

Comment:
it might be worth-while mentioning that we must prevent the reference to "auto_interrupt" from being finalised